File: android-native/app/src/main/java/com/example/kpkn/screens/wikilab/WikiLabExerciseSupport.kt
package com.example.kpkn.screens.wikilab

import androidx.compose.ui.graphics.Color
import com.example.kpkn.data.exercises.catalogv2.decodeCatalogRichMetadata
import com.example.kpkn.data.exercises.catalogExerciseIndex
import com.example.kpkn.data.exercises.resolveExercise
import com.example.kpkn.data.models.InvolvedMuscle
import com.example.kpkn.data.models.ExerciseMuscleInfo
import com.example.kpkn.data.models.MuscleRole
import com.example.kpkn.data.repository.WikiLabRepository
import com.example.kpkn.domain.exercises.catalogv2.AprendeOntology
import com.example.kpkn.domain.exercises.catalogv2.AprendeSimilarityBand
import com.example.kpkn.domain.exercises.catalogv2.AprendeSimilarityEngine
import com.example.kpkn.domain.exercises.catalogv2.AprendeSimilarityInput
import com.example.kpkn.domain.exercises.catalogv2.AprendeMuscleRole
import com.example.kpkn.domain.training.VolumeCalculator

internal data class WikiLabExerciseLink(
    val id: String,
    val name: String,
    val subtitle: String = "",
)

private val CANONICAL_MUSCLE_COLORS = mapOf(
    "Pectorales" to Color(0xFFE53935),
    "Dorsales" to Color(0xFF1E88E5),
    "Trapecio" to Color(0xFF1976D2),
    "Deltoides" to Color(0xFFFF8F00),
    "Tríceps" to Color(0xFF7B1FA2),
    "Bíceps" to Color(0xFF8E24AA),
    "Antebrazo" to Color(0xFF795548),
    "Abdomen" to Color(0xFF00897B),
    "Cuádriceps" to Color(0xFF43A047),
    "Isquiosurales" to Color(0xFF2E7D32),
    "Glúteos" to Color(0xFF558B2F),
    "Glúteo Medio" to Color(0xFF558B2F),
    "Aductores" to Color(0xFF7CB342),
    "Pantorrillas" to Color(0xFF33691E),
    "Core" to Color(0xFF00695C),
    "Erectores Espinales" to Color(0xFF1565C0),
    "Cuello" to Color(0xFF6D4C41),
)

internal fun wikilabMuscleColor(name: String): Color =
    CANONICAL_MUSCLE_COLORS[name] ?: Color(0xFF757575)

/** Local Aprende tokens. Anatomy is content, not a color legend. */
internal val APRENDE_BACKGROUND: Color = Color.Black
internal val APRENDE_PANEL: Color = Color(0xFF121212)
internal val APRENDE_PANEL_ELEVATED: Color = Color(0xFF181818)
internal val APRENDE_DIVIDER: Color = Color(0xFF2A2A2A)
internal val APRENDE_TEXT: Color = Color.White
internal val APRENDE_TEXT_SECONDARY: Color = Color.White.copy(alpha = 0.72f)
internal val APRENDE_TEXT_MUTED: Color = Color.White.copy(alpha = 0.5f)
internal val APRENDE_LINK_COLOR: Color = Color(0xFF9DB6C9)

/** Compatibility alias used by existing Learn surfaces. */
internal val APRENDE_MUTED_FILL: Color = APRENDE_PANEL

internal fun aprendeMuscleColor(@Suppress("UNUSED_PARAMETER") name: String): Color =
    Color(0xFF7F8D96)

internal fun resolveWikiLabExerciseLinks(
    ids: List<String>,
    subtitle: String = "",
): List<WikiLabExerciseLink> {
    // Static anatomy labels use only the versioned decisions in
    // AprendeOntology; an arbitrary display-name match is intentionally not
    // accepted here.
    val seen = linkedSetOf<String>()
    return ids.mapNotNull { requestedId ->
        val canonicalId = resolveExercise(requestedId)?.id
            ?: AprendeOntology.legacyExerciseId(requestedId)
            ?: return@mapNotNull null
        val exercise = resolveExercise(canonicalId) ?: return@mapNotNull null
        if (!seen.add(exercise.id)) return@mapNotNull null
        WikiLabExerciseLink(
            id = exercise.id,
            name = exercise.name,
            subtitle = subtitle,
        )
    }
}

internal fun resolveWikiLabMuscleId(muscleName: String): String? = when (muscleName) {
    "Pectorales" -> "pectoral"
    "Dorsales" -> "espalda"
    "Trapecio" -> "trapecio"
    "Deltoides" -> "deltoides"
    "Tríceps" -> "tríceps"
    "Bíceps" -> "bíceps"
    "Antebrazo" -> "antebrazo"
    "Abdomen" -> "abdomen"
    "Cuádriceps" -> "cuádriceps"
    "Isquiosurales" -> "isquiosurales"
    "Glúteos" -> "glúteos"
    "Aductores" -> "aductores"
    "Pantorrillas" -> "pantorrillas"
    "Core" -> "core"
    "Erectores Espinales" -> "erectores-espinales"
    "Cuello" -> "cuello"
    else -> null
}

internal fun canonicalMuscleDisplayName(raw: String, emphasis: String? = null): String =
    VolumeCalculator.normalizeCanonicalMuscleGroup(raw.replace('-', ' '), emphasis)

internal fun canonicalWikiLabMuscleId(raw: String, emphasis: String? = null): String? =
    resolveWikiLabMuscleId(canonicalMuscleDisplayName(raw, emphasis))

internal fun canonicalWikiLabMuscleIdFromCatalogId(catalogMuscleId: String): String? =
    AprendeOntology.wikiLabMuscleId(catalogMuscleId)

internal fun canonicalWikiLabPatternId(catalogPatternId: String): String? =
    AprendeOntology.wikiLabPatternId(catalogPatternId)

internal fun canonicalWikiLabMuscleIdFromEntityId(muscleId: String): String? {
    val entity = WikiLabRepository.getMuscleById(muscleId)
    val canonicalFromName = entity?.name?.let { canonicalWikiLabMuscleId(it) }
    if (canonicalFromName != null) return canonicalFromName
    val canonicalFromId = canonicalWikiLabMuscleId(muscleId)
    if (canonicalFromId != null) return canonicalFromId
    return entity?.let {
        if (
            it.id in setOf(
                "pectoral",
                "espalda",
                "hombros",
                "brazos",
                "piernas",
                "abdomen",
                "core",
                "deltoides",
                "bíceps",
                "tríceps",
                "antebrazo",
                "cuádriceps",
                "isquiosurales",
                "glúteos",
                "aductores",
                "pantorrillas",
                "erectores-espinales",
                "cuello",
                "trapecio",
            )
        ) it.id else null
    }
}

internal fun collapseInvolvedMusclesToCanonical(muscles: List<InvolvedMuscle>): List<InvolvedMuscle> {
    if (muscles.isEmpty()) return emptyList()

    val rolePriority = mapOf(
        MuscleRole.PRIMARY to 0,
        MuscleRole.SECONDARY to 1,
        MuscleRole.STABILIZER to 2,
        MuscleRole.NEUTRALIZER to 3,
    )

    val grouped = linkedMapOf<String, InvolvedMuscle>()
    muscles.forEach { item ->
        val canonical = canonicalMuscleDisplayName(item.muscle, item.emphasis)
        val existing = grouped[canonical]
        if (existing == null) {
            grouped[canonical] = item.copy(muscle = canonical)
        } else {
            val existingPriority = rolePriority[existing.role] ?: 99
            val incomingPriority = rolePriority[item.role] ?: 99
            val role = if (incomingPriority < existingPriority) item.role else existing.role
            val volumeContribution = maxOf(existing.volumeContribution ?: 0.0, item.volumeContribution ?: 0.0).takeIf { it > 0.0 }
            grouped[canonical] = existing.copy(muscle = canonical, role = role, volumeContribution = volumeContribution)
        }
    }
    return grouped.values.toList()
}

internal data class AprendeSimilarItem(
    val exercise: ExerciseMuscleInfo,
    val rationale: String,
    val score: Int,
)

internal data class AprendeExerciseRelations(
    val equivalent: List<AprendeSimilarItem> = emptyList(),
    val patternVariants: List<AprendeSimilarItem> = emptyList(),
    val anatomicalTransfer: List<AprendeSimilarItem> = emptyList(),
)

/**
 * Builds explicit relations from the resolved v2 profile. There is no
 * name-based or inferred setup fallback here. Custom exercises are kept in
 * the catalog display but do not receive synthetic v2 relations.
 */
internal fun buildAprendeExerciseRelations(
    info: ExerciseMuscleInfo,
    catalog: List<ExerciseMuscleInfo>,
    limit: Int = 4,
): AprendeExerciseRelations {
    val current = info.decodeCatalogRichMetadata() ?: return AprendeExerciseRelations()
    val currentInput = current.toAprendeSimilarityInput(info)
    val candidatePairs = catalog.asSequence()
        .filterNot { it.isCustom }
        .mapNotNull { candidate ->
            candidate.decodeCatalogRichMetadata()?.let { candidate to it.toAprendeSimilarityInput(candidate) }
        }
        // The shared index contains both the parent definition entry and the
        // explicit configuration entry for a default variant. They are the
        // same editorial record, so compare configuration identity rather
        // than allowing a relation list to show that record twice.
        .filterNot { (_, input) -> input.configurationId == current.identity.configurationId }
        .distinctBy { (_, input) -> input.configurationId ?: input.id }
        .toList()
    val candidateById = candidatePairs.associateBy { (candidate, _) -> candidate.id }
    val matches = AprendeSimilarityEngine.rank(
        current = currentInput,
        candidates = candidateById.values.map { (_, input) -> input },
        limit = maxOf(limit * 3, limit),
    )

    fun item(match: com.example.kpkn.domain.exercises.catalogv2.AprendeSimilarityMatch): AprendeSimilarItem {
        val candidate = candidateById.getValue(match.candidate.id).first
        val rationale = when {
            match.sameIntent -> "Misma intención de reemplazo: ${candidate.name}."
            match.sameDefinition -> "Otra configuración de la misma definición."
            match.samePattern && match.sharedJoints > 0 -> "Comparte patrón y articulaciones relevantes."
            match.sharedMuscles > 0 && match.sharedJoints > 0 -> "Comparte músculos y articulaciones relevantes."
            match.sharedMuscles > 0 -> "Comparte musculatura objetivo."
            else -> "Comparte el patrón de movimiento."
        }
        return AprendeSimilarItem(candidate, rationale, match.score)
    }

    fun top(band: AprendeSimilarityBand): List<AprendeSimilarItem> = matches
        .filter { it.band == band }
        .take(limit)
        .map(::item)

    return AprendeExerciseRelations(
        equivalent = top(AprendeSimilarityBand.EQUIVALENT),
        patternVariants = top(AprendeSimilarityBand.PATTERN_VARIANT),
        anatomicalTransfer = top(AprendeSimilarityBand.ANATOMICAL_TRANSFER),
    )
}

private fun com.example.kpkn.domain.exercises.catalogv2.ResolvedExerciseMetadataV2.toAprendeSimilarityInput(
    info: ExerciseMuscleInfo,
): AprendeSimilarityInput = AprendeSimilarityInput(
    id = info.id,
    displayName = info.name,
    definitionId = identity.definitionId,
    replacementGroup = replacement.replacementGroup,
    preservesIntent = replacement.preservesIntent.toSet(),
    movementPatternId = canonicalWikiLabPatternId(biomechanics.movementPatternId)
        ?: biomechanics.movementPatternId,
    muscles = (
        anatomy.primaryMuscles + anatomy.secondaryMuscles + anatomy.stabilizerMuscles
        ).toSet(),
    musclesByRole = mapOf(
        AprendeMuscleRole.PRIMARY to anatomy.primaryMuscles.toSet(),
        AprendeMuscleRole.SECONDARY to anatomy.secondaryMuscles.toSet(),
        AprendeMuscleRole.STABILIZER to anatomy.stabilizerMuscles.toSet(),
    ),
    joints = anatomy.jointInvolvement.map { it.jointId }.toSet(),
    bodyRegion = biomechanics.bodyRegion,
    kineticChain = biomechanics.kineticChain,
    laterality = biomechanics.laterality,
    equipmentId = biomechanics.equipmentId,
    configurationId = identity.configurationId,
)

/** Reverse links are derived from the current resolved catalog, never from
 * stale hand-authored example lists. */
internal fun catalogExercisesForMuscle(muscleId: String, limit: Int = 8): List<WikiLabExerciseLink> {
    val acceptedCatalogMuscleIds = AprendeOntology.catalogMuscleIdsForWikiLabEntity(muscleId)
    return approvedCatalogExerciseEntries()
        .filter { exercise ->
            if (exercise.isCustom) return@filter false
            val metadata = exercise.decodeCatalogRichMetadata() ?: return@filter false
            val ids = metadata.anatomy.primaryMuscles +
                metadata.anatomy.secondaryMuscles +
                metadata.anatomy.stabilizerMuscles
            ids.any { it in acceptedCatalogMuscleIds }
        }
        .sortedBy { it.name }
        .take(limit)
        .map { it.toWikiLabExerciseLink() }
}

internal fun catalogExercisesForJoint(jointId: String, limit: Int = 8): List<WikiLabExerciseLink> =
    approvedCatalogExerciseEntries()
        .filter { exercise ->
            exercise.decodeCatalogRichMetadata()?.anatomy?.jointInvolvement?.any { it.jointId == jointId } == true
        }
        .sortedBy { it.name }
        .take(limit)
        .map { it.toWikiLabExerciseLink() }

internal fun catalogExercisesForPattern(patternId: String, limit: Int = 8): List<WikiLabExerciseLink> =
    approvedCatalogExerciseEntries()
        .filter { exercise ->
            canonicalWikiLabPatternId(exercise.decodeCatalogRichMetadata()?.biomechanics?.movementPatternId.orEmpty()) == patternId
        }
        .sortedBy { it.name }
        .take(limit)
        .map { it.toWikiLabExerciseLink() }

private fun ExerciseMuscleInfo.toWikiLabExerciseLink(): WikiLabExerciseLink =
    WikiLabExerciseLink(
        id = id,
        name = name,
        subtitle = listOfNotNull(equipment, catalogVariantChips.takeIf { it.isNotEmpty() }?.joinToString(" · "))
            .joinToString(" · "),
    )

/** Full approved source for reverse links: defaults plus every explicit
 * configuration, with the default configuration deduplicated by identity. */
private fun approvedCatalogExerciseEntries(): List<ExerciseMuscleInfo> =
    catalogExerciseIndex()
        .values
        .asSequence()
        .filterNot { it.isCustom }
        .distinctBy { it.catalogConfigurationId ?: it.id }
        .toList()

File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Wikilab/WikiLabExerciseSupport.swift
import SwiftUI

internal struct WikiLabExerciseLink {
    let id: String
    let name: String
    let subtitle: String
    
    init(id: String, name: String, subtitle: String = "") {
        self.id = id
        self.name = name
        self.subtitle = subtitle
    }
}

private let CANONICAL_MUSCLE_COLORS: [String: Color] = [
    "Pectorales": Color(hex: 0xE53935),
    "Dorsales": Color(hex: 0x1E88E5),
    "Trapecio": Color(hex: 0x1976D2),
    "Deltoides": Color(hex: 0xFFFF8F00),
    "Tríceps": Color(hex: 0x7B1FA2),
    "Bíceps": Color(hex: 0x8E24AA),
    "Antebrazo": Color(hex: 0x795548),
    "Abdomen": Color(hex: 0x00897B),
    "Cuádriceps": Color(hex: 0x43A047),
    "Isquiosurales": Color(hex: 0x2E7D32),
    "Glúteos": Color(hex: 0x558B2F),
    "Aductores": Color(hex: 0x7CB342),
    "Pantorrillas": Color(hex: 0x33691E),
    "Core": Color(hex: 0x00695C),
    "Erectores Espinales": Color(hex: 0x1565C0),
    "Cuello": Color(hex: 0x6D4C41),
]

internal func wikilabMuscleColor(_ name: String) -> Color {
    CANONICAL_MUSCLE_COLORS[name] ?? Color(hex: 0x757575)
}

internal func resolveWikiLabExerciseLinks(ids: [String], subtitle: String = "") -> [WikiLabExerciseLink] {
    var seen = Set<String>()
    return ids.compactMap { requestedId in
        let exercise = resolveExercise(requestedId) ?? resolveExerciseFromNaturalLanguage(requestedId) ?? nil
        guard let ex = exercise else { return nil }
        guard !seen.contains(ex.id) else { return nil }
        seen.insert(ex.id)
        return WikiLabExerciseLink(id: ex.id, name: ex.name, subtitle: subtitle)
    }
}

private func resolveExerciseFromNaturalLanguage(_ raw: String) -> ExerciseMuscleInfo? {
    let query = normalizeForLookup(raw)
    guard !query.isEmpty,
          let hit = exerciseCatalogV2().search(query, limit: 1).first,
          let configuration = try? exerciseCatalogV2().defaultConfiguration(for: hit.definitionId)
    else { return nil }
    return resolveExercise(configuration.id)
}

private func normalizeForLookup(_ raw: String) -> String {
    raw.lowercased()
        .replacingOccurrences(of: "á", with: "a")
        .replacingOccurrences(of: "é", with: "e")
        .replacingOccurrences(of: "í", with: "i")
        .replacingOccurrences(of: "ó", with: "o")
        .replacingOccurrences(of: "ú", with: "u")
        .replacingOccurrences(of: "ü", with: "u")
        .replacingOccurrences(of: "ñ", with: "n")
        .replacingOccurrences(of: "-", with: " ")
        .replacingOccurrences(of: "_", with: " ")
        .trimmingCharacters(in: .whitespaces)
}

internal func resolveWikiLabMuscleId(_ muscleName: String) -> String? {
    switch muscleName {
    case "Pectorales": return "pectoral"
    case "Dorsales": return "espalda"
    case "Trapecio": return "trapecio"
    case "Deltoides": return "deltoides"
    case "Tríceps": return "tríceps"
    case "Bíceps": return "bíceps"
    case "Antebrazo": return "antebrazo"
    case "Abdomen": return "abdomen"
    case "Cuádriceps": return "cuádriceps"
    case "Isquiosurales": return "isquiosurales"
    case "Glúteos": return "glúteos"
    case "Aductores": return "aductores"
    case "Pantorrillas": return "pantorrillas"
    case "Core": return "core"
    case "Erectores Espinales": return "erectores-espinales"
    case "Cuello": return "cuello"
    default: return nil
    }
}

internal func canonicalMuscleDisplayName(_ raw: String, emphasis: String? = nil) -> String {
    normalizeCanonicalMuscleGroup(raw: raw.replacingOccurrences(of: "-", with: " "), emphasis: emphasis)
}

internal func canonicalWikiLabMuscleId(raw: String, emphasis: String? = nil) -> String? {
    resolveWikiLabMuscleId(canonicalMuscleDisplayName(raw, emphasis: emphasis))
}

internal func canonicalWikiLabMuscleIdFromEntityId(_ muscleId: String) -> String? {
    let entity = WikiLabRepository.shared.getMuscleById(id: muscleId)
    if let canonicalFromName = entity?.name.flatMap({ canonicalWikiLabMuscleId(raw: $0) }) {
        return canonicalFromName
    }
    if let canonicalFromId = canonicalWikiLabMuscleId(raw: muscleId) {
        return canonicalFromId
    }
    guard let ent = entity else { return nil }
    let knownIds: Set<String> = [
        "pectoral", "espalda", "hombros", "brazos", "piernas",
        "abdomen", "core", "deltoides", "bíceps", "tríceps",
        "antebrazo", "cuádriceps", "isquiosurales", "glúteos",
        "aductores", "pantorrillas", "erectores-espinales",
        "cuello", "trapecio",
    ]
    return knownIds.contains(ent.id) ? ent.id : nil
}

internal func collapseInvolvedMusclesToCanonical(_ muscles: [InvolvedMuscle]) -> [InvolvedMuscle] {
    guard !muscles.isEmpty else { return [] }
    
    let rolePriority: [MuscleRole: Int] = [
        .PRIMARY: 0,
        .SECONDARY: 1,
        .STABILIZER: 2,
        .NEUTRALIZER: 3,
    ]
    
    var grouped = [String: InvolvedMuscle]()
    for item in muscles {
        let canonical = canonicalMuscleDisplayName(item.muscle, emphasis: item.emphasis)
        if let existing = grouped[canonical] {
            let existingPrio = rolePriority[existing.role] ?? 99
            let incomingPrio = rolePriority[item.role] ?? 99
            let role = incomingPrio < existingPrio ? item.role : existing.role
            let vol: Double? = {
                let a = existing.volumeContribution ?? 0.0
                let b = item.volumeContribution ?? 0.0
                let m = max(a, b)
                return m > 0.0 ? m : nil
            }()
            grouped[canonical] = InvolvedMuscle(muscle: canonical, role: role, volumeContribution: vol, emphasis: nil)
        } else {
            grouped[canonical] = InvolvedMuscle(muscle: canonical, role: item.role, volumeContribution: item.volumeContribution, emphasis: nil)
        }
    }
    return Array(grouped.values)
}

internal func normalizeCanonicalMuscleGroup(raw: String, emphasis: String? = nil) -> String {
    let cleaned = raw.trimmingCharacters(in: .whitespaces)
    if cleaned.hasPrefix("Quad") || cleaned.hasPrefix("Cua") { return "Cuádriceps" }
    if cleaned.hasPrefix("Ham") || cleaned.hasPrefix("Isq") || cleaned.hasPrefix("Isquio") { return "Isquiosurales" }
    if cleaned.hasPrefix("Glute") || cleaned.hasPrefix("Glú") { return "Glúteos" }
    if cleaned.hasPrefix("Cal") || cleaned.hasPrefix("Pan") { return "Pantorrillas" }
    if cleaned.hasPrefix("Adductor") || cleaned.hasPrefix("Adu") { return "Aductores" }
    if cleaned.hasPrefix("Pec") { return "Pectorales" }
    if cleaned.hasPrefix("Lat") || cleaned.hasPrefix("Dor") { return "Dorsales" }
    if cleaned.hasPrefix("Trap") { return "Trapecio" }
    if cleaned.hasPrefix("Del") { return "Deltoides" }
    if cleaned.hasPrefix("Tri") { return "Tríceps" }
    if cleaned.hasPrefix("Bi") { return "Bíceps" }
    if cleaned.hasPrefix("Fore") || cleaned.hasPrefix("Ante") { return "Antebrazo" }
    if cleaned.hasPrefix("Spine") || cleaned.hasPrefix("Erector") || cleaned.hasPrefix("Erec") { return "Erectores Espinales" }
    if cleaned.hasPrefix("Neck") || cleaned.hasPrefix("Cue") { return "Cuello" }
    if cleaned.hasPrefix("Neck") || cleaned.hasPrefix("Abd") { return "Abdomen" }
    if cleaned == "core" || cleaned == "Core" { return "Core" }
    return cleaned
}

private func calculateSearchScore(_ exercise: ExerciseMuscleInfo, _ query: String) -> Int {
    let q = query.lowercased().trimmingCharacters(in: .whitespaces)
    if q.isEmpty { return 0 }
    let name = exercise.name.lowercased()
    var score = 0
    if name == q { score += 100 }
    else if name.hasPrefix(q) { score += 60 }
    else if name.contains(q) { score += 30 }
    
    let alias = exercise.alias?.lowercased() ?? ""
    if !alias.isEmpty {
        if alias == q { score += 80 }
        else if alias.hasPrefix(q) { score += 45 }
        else if alias.contains(q) { score += 20 }
    }
    
    for token in q.split(separator: " ") {
        if name.contains(token) { score += 12 }
        if alias.contains(token) { score += 8 }
    }
    
    let cat = exercise.category?.lowercased() ?? ""
    if cat.contains(q) { score += 10 }
    
    let equip = exercise.equipment?.lowercased() ?? ""
    if equip.contains(q) { score += 6 }
    
    return score
}

internal func deduplicateCatalogVisualResults(_ exercises: [ExerciseMuscleInfo]) -> [ExerciseMuscleInfo] {
    var seen = Set<String>()
    return exercises.filter { seen.insert($0.id.lowercased()).inserted }
}
